sql: TestGetUserHashedPasswordTimeout failed

[(sql).TestGetUserHashedPasswordTimeout failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2837627&tab=buildLog) on [master@6bc054abae5d7e177666e7cc3bc736bee3a99d8b](https://github.com/cockroachdb/cockroach/commits/6bc054abae5d7e177666e7cc3bc736bee3a99d8b):

```
=== RUN   TestGetUserHashedPasswordTimeout
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout832820028
    test_log_scope.go:74: use -show-logs to present logs inline
-- sanity checks --
-- make ranges unavailable --
-- expect timeout --
    user_test.go:154: expected error during connection, got ERROR: password authentication failed for user foo (SQLSTATE XXUUU)
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout832820028
--- FAIL: TestGetUserHashedPasswordTimeout (0.65s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestGetUserHashedPasswordTimeout PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestGetUserHashedPasswordTimeout.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestGetUserHashedPasswordTimeout failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2839856&tab=buildLog) on [master@1764c5f7c5642cff6dffe27432c90ab3a3484820](https://github.com/cockroachdb/cockroach/commits/1764c5f7c5642cff6dffe27432c90ab3a3484820):

```
=== RUN   TestGetUserHashedPasswordTimeout
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout422462952
    test_log_scope.go:74: use -show-logs to present logs inline
-- sanity checks --
-- make ranges unavailable --
-- expect timeout --
    user_test.go:154: expected error during connection, got ERROR: password authentication failed for user foo (SQLSTATE XXUUU)
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout422462952
--- FAIL: TestGetUserHashedPasswordTimeout (0.88s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestGetUserHashedPasswordTimeout PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestGetUserHashedPasswordTimeout.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestGetUserHashedPasswordTimeout failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2840837&tab=buildLog) on [master@a5f21439d2c455fa78028d54db2cccf6d6f6f7a2](https://github.com/cockroachdb/cockroach/commits/a5f21439d2c455fa78028d54db2cccf6d6f6f7a2):

```
=== RUN   TestGetUserHashedPasswordTimeout
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout392309448
    test_log_scope.go:74: use -show-logs to present logs inline
-- sanity checks --
-- make ranges unavailable --
-- expect timeout --
    user_test.go:154: expected error during connection, got ERROR: password authentication failed for user foo (SQLSTATE XXUUU)
E210401 21:55:31.040596 978422 kv/kvclient/kvcoord/txn_interceptor_committer.go:452  [-] 1  making txn commit explicit failed for "sql txn" meta={id=81a5fb8f pri=0.00233366 epo=0 ts=1617314130.823352485,0 min=1617314130.823352485,0 seq=6} lock=true stat=STAGING rts=1617314130.823352485,0 wto=false gul=1617314131.323352485,0 ifw=5: result is ambiguous (server shutdown)
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout392309448
--- FAIL: TestGetUserHashedPasswordTimeout (0.77s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestGetUserHashedPasswordTimeout PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestGetUserHashedPasswordTimeout.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestGetUserHashedPasswordTimeout failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2844878&tab=buildLog) on [master@df07592e26735be7362ef2b815ee163ead00739e](https://github.com/cockroachdb/cockroach/commits/df07592e26735be7362ef2b815ee163ead00739e):

```
=== RUN   TestGetUserHashedPasswordTimeout
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout418243074
    test_log_scope.go:74: use -show-logs to present logs inline
-- sanity checks --
-- make ranges unavailable --
-- expect timeout --
    user_test.go:154: expected error during connection, got ERROR: password authentication failed for user foo (SQLSTATE XXUUU)
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout418243074
--- FAIL: TestGetUserHashedPasswordTimeout (0.96s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestGetUserHashedPasswordTimeout PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestGetUserHashedPasswordTimeout.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestGetUserHashedPasswordTimeout failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2845451&tab=buildLog) on [master@df07592e26735be7362ef2b815ee163ead00739e](https://github.com/cockroachdb/cockroach/commits/df07592e26735be7362ef2b815ee163ead00739e):

```
=== RUN   TestGetUserHashedPasswordTimeout
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout768325144
    test_log_scope.go:74: use -show-logs to present logs inline
-- sanity checks --
-- make ranges unavailable --
-- expect timeout --
    user_test.go:154: expected error during connection, got ERROR: password authentication failed for user foo (SQLSTATE XXUUU)
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout768325144
--- FAIL: TestGetUserHashedPasswordTimeout (0.52s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestGetUserHashedPasswordTimeout PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestGetUserHashedPasswordTimeout.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestGetUserHashedPasswordTimeout failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2846153&tab=buildLog) on [master@df07592e26735be7362ef2b815ee163ead00739e](https://github.com/cockroachdb/cockroach/commits/df07592e26735be7362ef2b815ee163ead00739e):

```
=== RUN   TestGetUserHashedPasswordTimeout
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout036009719
    test_log_scope.go:74: use -show-logs to present logs inline
-- sanity checks --
-- make ranges unavailable --
-- expect timeout --
    user_test.go:154: expected error during connection, got ERROR: password authentication failed for user foo (SQLSTATE XXUUU)
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestGetUserHashedPasswordTimeout036009719
--- FAIL: TestGetUserHashedPasswordTimeout (0.91s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestGetUserHashedPasswordTimeout PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestGetUserHashedPasswordTimeout.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -738,10 +738,10 @@ func (r *DistSQLReceiver) Push(
 	}
 	r.tracing.TraceExecRowsResult(r.ctx, r.row)
 	if commErr := r.resultWriter.AddRow(r.ctx, r.row); commErr != nil {
-		if errors.Is(commErr, ErrLimitedResultClosed) {
-			// ErrLimitedResultClosed is not a real error, it is a signal to
-			// stop distsql and return success to the client (that's why we
-			// don't set the error on the resultWriter).
+		if errors.Is(commErr, ErrLimitedResultClosed) || errors.Is(commErr, errIEResultChannelClosed) {
+			// ErrLimitedResultClosed and errIEResultChannelClosed are not real
+			// errors, it is a signal to stop distsql and return success to the
+			// client (that's why we don't set the error on the resultWriter).
 			r.status = execinfra.DrainRequested
 		} else {
 			// Set the error on the resultWriter to notify the consumer about

--- a/pkg/sql/internal.go
+++ b/pkg/sql/internal.go
@@ -604,7 +604,7 @@ var rowsAffectedResultColumns = colinfo.ResultColumns{
 func (ie *InternalExecutor) execInternal(
 	ctx context.Context,
 	opName string,
-	rw ieResultChannel,
+	rw *ieResultChannel,
 	txn *kv.Txn,
 	sessionDataOverride sessiondata.InternalExecutorOverride,
 	stmt string,

--- a/pkg/sql/internal_result_channel.go
+++ b/pkg/sql/internal_result_channel.go
@@ -18,13 +18,6 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
-// ieResultChannel is used to coordinate passing results from an
-// internalExecutor to its corresponding iterator.
-type ieResultChannel interface {
-	ieResultReader
-	ieResultWriter
-}
-
 // ieResultReader is used to read internalExecutor results.
 // It is managed by the rowsIterator.
 type ieResultReader interface {
@@ -34,15 +27,13 @@ type ieResultReader interface {
 	// currently blocked and waits for the initial result to be written.
 	firstResult(ctx context.Context) (_ ieIteratorResult, done bool, err error)
 
-	// nextResult returns the nextResult. Done will always be true if err
+	// nextResult returns the next result. Done will always be true if err
 	// is non-nil. Err will be non-nil if either close has been called or
 	// the passed context is finished.
 	nextResult(ctx context.Context) (_ ieIteratorResult, done bool, err error)
 
-	// close ensures that the either writer has finished writing. In the case
-	// of an asynchronous channel, close will drain the writer's channel. In the
-	// case of the synchronous channel, it will ensure that the writer receives
-	// an error when it wakes.
+	// close ensures that either writer has finished writing. The writer will
+	// receive a signal to drain, and close will drain the writer's channel.
 	close() error
 }
 
@@ -67,174 +58,161 @@ var asyncIEResultChannelBufferSize = util.ConstantWithMetamorphicTestRange(
 
 // newAsyncIEResultChannel returns an ieResultChannel which does not attempt to
 // synchronize the writer with the reader.
-func newAsyncIEResultChannel() ieResultChannel {
-	return &asyncIEResultChannel{
+func newAsyncIEResultChannel() *ieResultChannel {
+	return &ieResultChannel{
 		dataCh: make(chan ieIteratorResult, asyncIEResultChannelBufferSize),
+		doneCh: make(chan struct{}),
 	}
 }
 
-type asyncIEResultChannel struct {
-	dataCh chan ieIteratorResult
-}
-
-var _ ieResultChannel = &asyncIEResultChannel{}
-
-func (c *asyncIEResultChannel) firstResult(
-	ctx context.Context,
-) (_ ieIteratorResult, done bool, err error) {
-	select {
-	case <-ctx.Done():
-		return ieIteratorResult{}, true, ctx.Err()
-	case res, ok := <-c.dataCh:
-		if !ok {
-			return ieIteratorResult{}, true, nil
-		}
-		return res, false, nil
-	}
-}
-
-func (c *asyncIEResultChannel) nextResult(
-	ctx context.Context,
-) (_ ieIteratorResult, done bool, err error) {
-	return c.firstResult(ctx)
-}
-
-func (c *asyncIEResultChannel) close() error {
-	var firstErr error
-	for {
-		res, done, err := c.nextResult(context.TODO())
-		if firstErr == nil {
-			if res.err != nil {
-				firstErr = res.err
-			} else if err != nil {
-				firstErr = err
-			}
-		}
-		if done {
-			return firstErr
-		}
-	}
-}
-
-func (c *asyncIEResultChannel) addResult(ctx context.Context, result ieIteratorResult) error {
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case c.dataCh <- result:
-		return nil
-	}
-}
-
-func (c *asyncIEResultChannel) finish() {
-	close(c.dataCh)
-}
-
-// syncIEResultChannel is used to ensure that in execution scenarios which
-// do not permit concurrency that there is none. It works by blocking the
-// writing goroutine immediately upon sending on the data channel and only
-// unblocking it after the reader signals.
-type syncIEResultChannel struct {
+// ieResultChannel is used to coordinate passing results from an
+// internalExecutor to its corresponding iterator. It can be constructed to
+// ensure that there is no concurrency between the reader and writer.
+type ieResultChannel struct {
 
 	// dataCh is the channel on which the connExecutor goroutine sends the rows
-	// (in addResult) and will block on waitCh after each send. The iterator
-	// goroutine blocks on dataCh until there is something to receive (rows or
-	// other metadata) and will return the data to the caller. On the next call
-	// to Next(), the iterator goroutine unblocks the producer and will block
-	// itself again. dataCh will be closed (in finish()) when the connExecutor
-	// goroutine exits its run() loop whereas waitCh is closed when closing the
-	// iterator.
+	// (in addResult) and, in the synchronous case, will block on waitCh after
+	// each send. The iterator goroutine blocks on dataCh until there is
+	// something to receive (rows or other metadata) and will return the data to
+	// the caller. On the next call to Next(), the iterator goroutine unblocks
+	// the producer and will block itself again. dataCh will be closed (in
+	// finish()) when the connExecutor goroutine exits its run() loop whereas
+	// waitCh is closed when closing the iterator.
 	dataCh chan ieIteratorResult
 
-	// waitCh is never closed. In all places where the caller may interact with it
-	// the doneCh is also used. This policy is in place to make it safe to unblock
-	// both the reader and the writer without any hazards of a blocked reader
-	// attempting to send on a closed channel.
+	// waitCh is nil for async ieResultChannels. It is never closed. In all places
+	// where the caller may interact with it the doneCh is also used. This policy
+	// is in place to make it safe to unblock both the reader and the writer
+	// without any hazards of a blocked reader attempting to send on a closed
+	// channel.
 	waitCh chan struct{}
 
-	// doneCh is used to indicate that the ReadWriter has been closed.
-	// doneCh is closed under the doneOnce. The doneCh is only used for the
-	// syncIEResultChannel. This is crucial to ensure that a synchronous writer
-	// does not attempt to continue to operate after the reader has called close.
+	// doneCh is used to indicate that the ieResultReader has been closed and is
+	// closed under the doneOnce, the writer will transition to draining. This
+	// is crucial to ensure that a synchronous writer does not attempt to
+	// continue to operate after the reader has called close.
 	doneCh   chan struct{}
+	doneErr  error
 	doneOnce sync.Once
 }
 
-var _ ieResultChannel = &syncIEResultChannel{}
-
-// newSyncIEResultChannel returns an ieResultChannel which synchronizes the
-// writer with the reader.
-func newSyncIEResultChannel() ieResultChannel {
-	return &syncIEResultChannel{
+// newSyncIEResultChannel is used to ensure that in execution scenarios which
+// do not permit concurrency that there is none. It works by blocking the
+// writing goroutine immediately upon sending on the data channel and only
+// unblocking it after the reader signals.
+func newSyncIEResultChannel() *ieResultChannel {
+	return &ieResultChannel{
 		dataCh: make(chan ieIteratorResult),
 		waitCh: make(chan struct{}),
 		doneCh: make(chan struct{}),
 	}
 }
 
-func (i *syncIEResultChannel) firstResult(
+func (i *ieResultChannel) firstResult(
 	ctx context.Context,
 ) (_ ieIteratorResult, done bool, err error) {
 	select {
 	case <-ctx.Done():
 		return ieIteratorResult{}, true, ctx.Err()
 	case <-i.doneCh:
-		return ieIteratorResult{}, true, nil
+		return ieIteratorResult{}, true, ctx.Err()
 	case res, ok := <-i.dataCh:
 		if !ok {
-			return ieIteratorResult{}, true, nil
+			return ieIteratorResult{}, true, ctx.Err()
 		}
 		return res, false, nil
 	}
 }
 
-func (i *syncIEResultChannel) unblockWriter(ctx context.Context) (done bool, err error) {
+func (i *ieResultChannel) maybeUnblockWriter(ctx context.Context) (done bool, err error) {
+	if i.async() {
+		return false, nil
+	}
 	select {
 	case <-ctx.Done():
 		return true, ctx.Err()
 	case <-i.doneCh:
-		return true, nil
+		return true, ctx.Err()
 	case i.waitCh <- struct{}{}:
 		return false, nil
 	}
 }
 
-func (i *syncIEResultChannel) finish() {
-	close(i.dataCh)
+func (i *ieResultChannel) async() bool {
+	return i.waitCh == nil
 }
 
-func (i *syncIEResultChannel) nextResult(
+func (i *ieResultChannel) nextResult(
 	ctx context.Context,
 ) (_ ieIteratorResult, done bool, err error) {
-	if done, err = i.unblockWriter(ctx); done {
+	if done, err = i.maybeUnblockWriter(ctx); done {
 		return ieIteratorResult{}, done, err
 	}
 	return i.firstResult(ctx)
 }
 
-func (i *syncIEResultChannel) close() error {
-	i.doneOnce.Do(func() { close(i.doneCh) })
-	return nil
+func (i *ieResultChannel) close() error {
+	i.doneOnce.Do(func() {
+		close(i.doneCh)
+		for {
+			// In the async case, res might contain some actual rows, but we're
+			// not interested in them; in the sync case, only errors are
+			// expected to be retrieved from now one because the writer
+			// transitions to draining.
+			res, done, err := i.nextResult(context.TODO())
+			if i.doneErr == nil {
+				if res.err != nil {
+					i.doneErr = res.err
+				} else if err != nil {
+					i.doneErr = err
+				}
+			}
+			if done {
+				return
+			}
+		}
+	})
+	return i.doneErr
 }
 
-// errSyncIEResultReaderCanceled is returned by the writer when the reader has
-// closed syncIEResultChannel. The error indicates to the writer to shut down
-// the query execution, but the reader won't propagate it further.
-var errSyncIEResultReaderCanceled = errors.New("synchronous ieResultReader closed")
+// errIEResultChannelClosed is returned by the writer when the reader has closed
+// ieResultChannel. The error indicates to the writer to drain the query
+// execution, but the reader won't propagate it further.
+var errIEResultChannelClosed = errors.New("ieResultReader closed")
 
-func (i *syncIEResultChannel) addResult(ctx context.Context, result ieIteratorResult) error {
+func (i *ieResultChannel) addResult(ctx context.Context, result ieIteratorResult) error {
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	case <-i.doneCh:
-		return errSyncIEResultReaderCanceled
+		// Prefer the context error if there is one.
+		if ctxErr := ctx.Err(); ctxErr != nil {
+			return ctxErr
+		}
+		return errIEResultChannelClosed
 	case i.dataCh <- result:
+	}
+	return i.maybeBlock(ctx)
+}
+
+func (i *ieResultChannel) maybeBlock(ctx context.Context) error {
+	if i.async() {
+		return nil
 	}
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	case <-i.doneCh:
-		return errSyncIEResultReaderCanceled
+		// Prefer the context error if there is one.
+		if ctxErr := ctx.Err(); ctxErr != nil {
+			return ctxErr
+		}
+		return errIEResultChannelClosed
 	case <-i.waitCh:
 		return nil
 	}
+}
+
+func (i *ieResultChannel) finish() {
+	close(i.dataCh)
 }

--- a/pkg/sql/user_test.go
+++ b/pkg/sql/user_test.go
@@ -45,8 +45,6 @@ func TestGetUserHashedPasswordTimeout(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 62948 /* githubIssueID */)
-
 	// We want to use a low timeout below to prevent
 	// this test from taking forever, however
 	// race builds are so slow as to trigger this timeout spuriously.

